From 3ed6b636398104eb7533cb572917e69b34c60a81 Mon Sep 17 00:00:00 2001
From: Robert Nelson <robertcnelson@gmail.com>
Date: Thu, 17 Sep 2015 15:44:31 -0500
Subject: [PATCH 53/97] apply backport patch
 collateral-evolutions/network/0038-ethtool_rxfh_indir/igb_rxfh_indir.patch

---
 .../drivers/net/ethernet/intel/igb/igb_ethtool.c   | 27 ++++++++++++++++++++++
 1 file changed, 27 insertions(+)

diff --git a/backports/drivers/net/ethernet/intel/igb/igb_ethtool.c b/backports/drivers/net/ethernet/intel/igb/igb_ethtool.c
index d5673eb..1946878 100644
--- a/backports/drivers/net/ethernet/intel/igb/igb_ethtool.c
+++ b/backports/drivers/net/ethernet/intel/igb/igb_ethtool.c
@@ -2837,19 +2837,28 @@ static void igb_ethtool_complete(struct net_device *netdev)
 	pm_runtime_put(&adapter->pdev->dev);
 }
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,3,0)
 static u32 igb_get_rxfh_indir_size(struct net_device *netdev)
 {
 	return IGB_RETA_SIZE;
 }
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,19,0)
 static int igb_get_rxfh(struct net_device *netdev, u32 *indir, u8 *key,
 			u8 *hfunc)
+#elif LINUX_VERSION_CODE >= KERNEL_VERSION(3,16,0)
+static int igb_get_rxfh(struct net_device *netdev, u32 *indir, u8 *key)
+#else
+static int igb_get_rxfh(struct net_device *netdev, u32 *indir)
+#endif
 {
 	struct igb_adapter *adapter = netdev_priv(netdev);
 	int i;
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,19,0)
 	if (hfunc)
 		*hfunc = ETH_RSS_HASH_TOP;
+#endif
 	if (!indir)
 		return 0;
 	for (i = 0; i < IGB_RETA_SIZE; i++)
@@ -2857,6 +2866,7 @@ static int igb_get_rxfh(struct net_device *netdev, u32 *indir, u8 *key,
 
 	return 0;
 }
+#endif /* LINUX_VERSION_CODE >= KERNEL_VERSION(3,3,0) */
 
 void igb_write_rss_indir_tbl(struct igb_adapter *adapter)
 {
@@ -2893,18 +2903,28 @@ void igb_write_rss_indir_tbl(struct igb_adapter *adapter)
 	}
 }
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,3,0)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,19,0)
 static int igb_set_rxfh(struct net_device *netdev, const u32 *indir,
 			const u8 *key, const u8 hfunc)
+#elif LINUX_VERSION_CODE >= KERNEL_VERSION(3,16,0)
+static int igb_set_rxfh(struct net_device *netdev, const u32 *indir,
+			const u8 *key)
+#else
+static int igb_set_rxfh(struct net_device *netdev, const u32 *indir)
+#endif
 {
 	struct igb_adapter *adapter = netdev_priv(netdev);
 	struct e1000_hw *hw = &adapter->hw;
 	int i;
 	u32 num_queues;
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,19,0)
 	/* We do not allow change in unsupported parameters */
 	if (key ||
 	    (hfunc != ETH_RSS_HASH_NO_CHANGE && hfunc != ETH_RSS_HASH_TOP))
 		return -EOPNOTSUPP;
+#endif
 	if (!indir)
 		return 0;
 
@@ -2933,6 +2953,7 @@ static int igb_set_rxfh(struct net_device *netdev, const u32 *indir,
 
 	return 0;
 }
+#endif /* LINUX_VERSION_CODE >= KERNEL_VERSION(3,3,0) */
 
 static unsigned int igb_max_channels(struct igb_adapter *adapter)
 {
@@ -3049,9 +3070,15 @@ static const struct ethtool_ops igb_ethtool_ops = {
 	.set_eee		= igb_set_eee,
 	.get_module_info	= igb_get_module_info,
 	.get_module_eeprom	= igb_get_module_eeprom,
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,16,0)
 	.get_rxfh_indir_size	= igb_get_rxfh_indir_size,
 	.get_rxfh		= igb_get_rxfh,
 	.set_rxfh		= igb_set_rxfh,
+#elif LINUX_VERSION_CODE >= KERNEL_VERSION(3,3,0)
+	.get_rxfh_indir_size	= igb_get_rxfh_indir_size,
+	.get_rxfh_indir		= igb_get_rxfh,
+	.set_rxfh_indir		= igb_set_rxfh,
+#endif /* LINUX_VERSION_CODE >= KERNEL_VERSION(3,3,0) */
 	.get_channels		= igb_get_channels,
 	.set_channels		= igb_set_channels,
 	.begin			= igb_ethtool_begin,
-- 
2.5.1

